Sources/StreamVideo/Utils/AudioRecorder/StreamAudioRecorder.swift

```diff
@@ -0,0 +1,180 @@
+
+    /// A public publisher that exposes the average power of the audio signal.
+    open private(set) lazy var metersPublisher: AnyPublisher<Float, Never> = _metersPublisher.eraseToAnyPublisher()
+
+    /// Indicates whether an active call is present, influencing recording behavior.
+    public var hasActiveCall: Bool = false {
+        didSet {
+            if !hasActiveCall {
+                Task {
+                    await stopRecording()
+                    try? audioSession.setActive(false, options: [])
+                }
+            }
+        }
+    }
+
+    /// Initializes the recorder with a filename.
+    ///
+    /// - Parameter filename: The name of the file to record to.
+    public init(filename: String) {
+        audioRecorderBuilder = .init(inCacheDirectoryWithFilename: filename)
+        audioSession = AVAudioSession.sharedInstance()
+
+        setUp()
+    }
+
+    /// Initializes the recorder with a custom builder and audio session.
+    ///
+    /// - Parameter audioRecorderBuilder: The builder used to create the recorder.
+    /// - Parameter audioSession: The audio session used for recording and playback.
+    init(
+        audioRecorderBuilder: AVAudioRecorderBuilder,
+        audioSession: AudioSessionProtocol
+    ) {
+        self.audioRecorderBuilder = audioRecorderBuilder
+        self.audioSession = audioSession
+
+        setUp()
+    }
+
+    deinit {
+        setUpTask?.cancel()
+        setUpTask = nil
+        do {
+            try FileManager.default.removeItem(at: audioRecorderBuilder.fileURL)
+            log.debug("Successfully deleted audio filename")
+        } catch {
+            log.error("Error deleting fileURL.", error: error)
+        }
+    }
+
+    // MARK: - Public API
+
+    /// Starts recording audio asynchronously.
+    open func startRecording() async {
+        await setUpAudioCaptureIfRequired { [weak self] audioRecorder in
+            guard let self, self.hasActiveCall, !audioRecorder.isRecording else {
+                return
+            }
+
+            audioRecorder.record()
+            audioRecorder.isMeteringEnabled = true
+
+            log.debug("️Recording started.")
+            self.updateMetersTimerCancellable = Foundation.Timer
+                .publish(every: 0.1, on: .main, in: .default)
+                .autoconnect()
+                .sink { [weak self, audioRecorder] _ in
+                    Task { [weak self, audioRecorder] in
+                        guard let self else { return }
+                        audioRecorder.updateMeters()
+                        self._metersPublisher.send(audioRecorder.averagePower(forChannel: 0))
+                        log.debug("️Recording meters updated")
+                    }
+                }
+        }
+    }
+
+    /// Stops recording audio asynchronously.
+    open func stopRecording() async {
+        guard let audioRecorder = await audioRecorderBuilder.result, audioRecorder.isRecording else {
```

Comment:
thinking if we should do some cleanup of the file also when we stop recording (not only on deinit), wdyt?

Reply:
It won't hurt for sure